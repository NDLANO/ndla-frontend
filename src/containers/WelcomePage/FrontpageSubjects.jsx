/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';

import { FrontpageSubjects as FrontpageSubjectsSection } from 'ndla-ui';
import { toSubject } from '../../routeHelpers';
import {
  GraphQLFrontpageCategoryShape,
  GraphQLSimpleSubjectShape,
} from '../../graphqlShapes';
import config from '../../config';
import { OLD_CATEGORIES_WITH_SUBJECTS } from '../../constants';

export const getAllImportSubjectsCategory = (subjects = []) => ({
  imported: {
    name: 'imported',
    subjects: subjects.map(subject => ({
      text: subject.name,
      url: toSubject(subject.id),
      yearInfo: subject.yearInfo,
    })),
  },
});

const sortByName = arr =>
  arr.sort((a, b) => {
    if (a.name < b.name) return -1;
    if (a.name > b.name) return 1;
    return 0;
  });
export const getCategoriesWithAllSubjects = (
  categoriesFromApi = [],
  locale,
) => {
  // en is only valid for english nodes in old system
  const lang = locale === 'en' ? 'nb' : locale;
  return categoriesFromApi.map(category => {
    const newSubjects = category.subjects.map(categorySubject => ({
      ...categorySubject,
      text: categorySubject.name,
      url: toSubject(categorySubject.id),
      yearInfo: categorySubject.yearInfo,
    }));
    const oldSubjects = OLD_CATEGORIES_WITH_SUBJECTS[category.name]
      .map(subject => ({
        ...subject,
        id: subject.nodeId,
        text: subject.name,
        url: subject.lang
          ? `/${subject.lang}/node/${subject.nodeId}`
          : `/${lang}/node/${subject.nodeId}`,
        yearInfo: subject.yearInfo,
      }))
      .filter(
        oldSubject =>
          newSubjects.find(newSubject =>
            oldSubject.name.startsWith(newSubject.name),
          ) === undefined,
      );
    return {
      [category.name]: {
        ...category,
        subjects: sortByName([...oldSubjects, ...newSubjects]),
      },
    };
  });
};

const FrontpageSubjects = ({ categories, subjects, locale }) => {
  const frontpageCategories = getCategoriesWithAllSubjects(
    categories,
    locale,
  ).reduce((obj, item) => ({
    ...obj,
    [obj[Object.keys(item)]]: item[Object.keys(item)],
  }));

  const allSubjects = config.isNdlaProdEnvironment
    ? { ...frontpageCategories }
    : {
        ...frontpageCategories,
        ...getAllImportSubjectsCategory(subjects),
      };

  return <FrontpageSubjectsSection subjects={allSubjects} />;
};

FrontpageSubjects.propTypes = {
  locale: PropTypes.string.isRequired,
  categories: PropTypes.arrayOf(GraphQLFrontpageCategoryShape),
  subjects: PropTypes.arrayOf(GraphQLSimpleSubjectShape),
};

FrontpageSubjects.defaultProps = {
  categories: [],
};

export default FrontpageSubjects;
