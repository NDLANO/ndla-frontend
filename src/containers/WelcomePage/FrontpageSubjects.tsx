/**
/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { FrontpageProgramMenu } from '@ndla/ui';

import {
  getCategorizedSubjects,
  getProgrammes,
} from '../../util/programmesSubjectsHelper';
import { LocaleType } from '../../interfaces';

interface Props {
  locale: LocaleType;
}

const FrontpageSubjects = ({ locale }: Props) => (
  <FrontpageProgramMenu
    programItems={getProgrammes(locale)}
    subjectCategories={getCategorizedSubjects(locale)}
  />
);

FrontpageSubjects.propTypes = {
  locale: PropTypes.string.isRequired,
};

export default FrontpageSubjects;
