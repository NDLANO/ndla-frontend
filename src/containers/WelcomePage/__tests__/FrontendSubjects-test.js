/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { mapHardCodedCategories } from '../FrontpageSubjects';

const categoriesFromApi = [
  {
    name: 'fellesfag',
    subjects: [],
  },
  {
    name: 'yrkesfag',
    subjects: [],
  },
  {
    name: 'studiespesialiserende',
    subjects: [],
  },
];

const specializationSubjects = [
  {
    id: 'urn:subject:7',
    name: 'MarkedsfÃ¸ring og ledelse 1',
    path: '/subject:7',
    frontpageFilters: [],
  },
  {
    id: 'urn:subject:2',
    name: 'Kinesisk',
    frontpageFilters: [
      { name: 'Kinesisk 1', id: 'urn:filter:1337' },
      { name: 'Kinesisk 2', id: 'urn:filter:42' },
    ],
    path: '/subject:2',
  },
  {
    id: 'urn:subject:1',
    name: 'Medieuttrykk og mediesamfunnet',
    path: '/subject:1',
    frontpageFilters: [],
  },
];

const vocationalSubjects = [
  {
    id: 'urn:subject:12',
    name: 'Service og samferdsel Vg1',
    path: '/subject:12',
    frontpageFilters: [],
  },
];

test('mapHardCodedCategories all old node subjects (nb locale)', () => {
  const categories = mapHardCodedCategories(categoriesFromApi, 'nb');

  expect(categories).toMatchSnapshot();
});

test('mapHardCodedCategories with all old node subjects (nn locale)', () => {
  const categories = mapHardCodedCategories(categoriesFromApi, 'nn');

  expect(categories).toMatchSnapshot();
});

test('mapHardCodedCategories with all old node subjects (en locale)', () => {
  const categories = mapHardCodedCategories(categoriesFromApi, 'en');

  expect(categories).toMatchSnapshot();
});

test('mapHardCodedCategories with some specialization subjects replaced', () => {
  const specializationCategory = {
    ...categoriesFromApi[2],
    subjects: specializationSubjects,
  };
  const categories = mapHardCodedCategories([specializationCategory], 'nb');

  expect(categories).toMatchSnapshot();
});

test('mapHardCodedCategories with a vocational subjects replaced', () => {
  const vocationalCategory = {
    ...categoriesFromApi[1],
    subjects: vocationalSubjects,
  };
  const categories = mapHardCodedCategories([vocationalCategory], 'nb');

  expect(categories).toMatchSnapshot();
});
