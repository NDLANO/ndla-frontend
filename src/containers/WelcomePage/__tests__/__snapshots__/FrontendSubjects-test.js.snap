// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapHardCodedCategories without frontpageFilters from api (nb locale) 1`] = `
Array [
  Object {
    "name": "fellesfag",
    "subjects": Array [
      Object {
        "id": "urn:subject:39",
        "lang": "en",
        "name": "Engelsk",
        "nodeId": "42",
        "text": "Engelsk",
        "url": "/subjects/subject:39",
      },
      Object {
        "id": "urn:subject:9",
        "name": "Historie Vg2 og Vg3",
        "nodeId": "52253",
        "text": "Historie Vg2 og Vg3",
        "url": "/subjects/subject:9",
      },
      Object {
        "id": "urn:subject:26",
        "name": "Kroppsøving",
        "nodeId": "46",
        "text": "Kroppsøving",
        "url": "/subjects/subject:26",
      },
      Object {
        "id": "urn:subject:34",
        "name": "Matematikk 1P",
        "nodeId": "55",
        "text": "Matematikk 1P",
        "url": "/subjects/subject:34",
      },
      Object {
        "id": "urn:subject:30",
        "name": "Matematikk 1P-Y",
        "text": "Matematikk 1P-Y",
        "url": "/subjects/subject:30",
      },
      Object {
        "id": "urn:subject:33",
        "name": "Matematikk 1T",
        "nodeId": "54",
        "text": "Matematikk 1T",
        "url": "/subjects/subject:33",
      },
      Object {
        "id": "urn:subject:30",
        "name": "Matematikk 1T-Y",
        "text": "Matematikk 1T-Y",
        "url": "/subjects/subject:30",
      },
      Object {
        "id": "urn:subject:29",
        "name": "Matematikk 2P",
        "nodeId": "53",
        "text": "Matematikk 2P",
        "url": "/subjects/subject:29",
      },
      Object {
        "id": "urn:subject:30",
        "name": "Matematikk 2P-Y",
        "text": "Matematikk 2P-Y",
        "url": "/subjects/subject:30",
      },
      Object {
        "id": "urn:subject:21",
        "name": "Naturfag",
        "nodeId": "7",
        "text": "Naturfag",
        "url": "/subjects/subject:21",
      },
      Object {
        "id": "urn:subject:19",
        "name": "Norsk Vg2 og Vg3 SF",
        "nodeId": "27",
        "shortname": "SF Vg2,SF Vg3",
        "text": "Norsk Vg2 og Vg3 SF",
        "url": "/subjects/subject:19",
      },
      Object {
        "id": "urn:subject:19",
        "name": "Norsk YF og SF",
        "nodeId": "116784",
        "shortname": "SF Vg1,PB Vg3,YF Vg1/Vg2",
        "text": "Norsk YF og SF",
        "url": "/subjects/subject:19",
      },
      Object {
        "beta": true,
        "id": "urn:subject:44",
        "name": "Religion og etikk",
        "nodeId": "185100",
        "text": "Religion og etikk",
        "url": "/subjects/subject:44",
      },
      Object {
        "id": "urn:subject:3",
        "name": "Samfunnsfag",
        "nodeId": "36",
        "text": "Samfunnsfag",
        "url": "/subjects/subject:3",
      },
      Object {
        "beta": true,
        "id": "urn:subject:15",
        "name": "Sørsamisk som førstespråk",
        "nodeId": "126960",
        "text": "Sørsamisk som førstespråk",
        "url": "/subjects/subject:15",
      },
    ],
  },
  Object {
    "name": "yrkesfag",
    "subjects": Array [
      Object {
        "id": "urn:subject:40",
        "name": "Barne- og ungdomsarbeiderfag Vg2",
        "nodeId": "51",
        "text": "Barne- og ungdomsarbeiderfag Vg2",
        "url": "/subjects/subject:40",
      },
      Object {
        "id": "urn:subject:6",
        "name": "Brønnteknikk",
        "nodeId": "137414",
        "text": "Brønnteknikk",
        "url": "/subjects/subject:6",
      },
      Object {
        "id": "urn:subject:11",
        "name": "Bygg- og anleggsteknikk Vg1",
        "nodeId": "127013",
        "text": "Bygg- og anleggsteknikk Vg1",
        "url": "/subjects/subject:11",
      },
      Object {
        "id": "urn:subject:38",
        "name": "Design og håndverk Vg1",
        "nodeId": "44",
        "text": "Design og håndverk Vg1",
        "url": "/subjects/subject:38",
      },
      Object {
        "id": "urn:subject:16",
        "name": "Elektrofag Vg1",
        "nodeId": "43",
        "text": "Elektrofag Vg1",
        "url": "/subjects/subject:16",
      },
      Object {
        "id": "urn:subject:24",
        "name": "Helse- og oppvekstfag Vg1",
        "nodeId": "8",
        "text": "Helse- og oppvekstfag Vg1",
        "url": "/subjects/subject:24",
      },
      Object {
        "id": "urn:subject:4",
        "name": "Helsearbeiderfag Vg2",
        "nodeId": "52",
        "text": "Helsearbeiderfag Vg2",
        "url": "/subjects/subject:4",
      },
      Object {
        "id": "urn:subject:25",
        "name": "IKT-servicefag Vg2",
        "nodeId": "102783",
        "text": "IKT-servicefag Vg2",
        "url": "/subjects/subject:25",
      },
      Object {
        "id": "urn:subject:41",
        "name": "Kokk- og servitørfag Vg2",
        "nodeId": "86643",
        "text": "Kokk- og servitørfag Vg2",
        "url": "/subjects/subject:41",
      },
      Object {
        "id": "urn:subject:13",
        "name": "Naturbruk Vg1",
        "nodeId": "137415",
        "text": "Naturbruk Vg1",
        "url": "/subjects/subject:13",
      },
      Object {
        "id": "urn:subject:35",
        "name": "Reiseliv Vg2",
        "nodeId": "102780",
        "text": "Reiseliv Vg2",
        "url": "/subjects/subject:35",
      },
      Object {
        "id": "urn:subject:37",
        "name": "Restaurant- og matfag Vg1",
        "nodeId": "37",
        "text": "Restaurant- og matfag Vg1",
        "url": "/subjects/subject:37",
      },
      Object {
        "id": "urn:subject:22",
        "name": "Salg, service og sikkerhet Vg2",
        "nodeId": "102781",
        "text": "Salg, service og sikkerhet Vg2",
        "url": "/subjects/subject:22",
      },
      Object {
        "id": "urn:subject:12",
        "name": "Service og samferdsel Vg1",
        "nodeId": "52291",
        "text": "Service og samferdsel Vg1",
        "url": "/subjects/subject:12",
      },
      Object {
        "id": "urn:subject:28",
        "name": "Teknikk og industriell produksjon Vg1",
        "nodeId": "35",
        "text": "Teknikk og industriell produksjon Vg1",
        "url": "/subjects/subject:28",
      },
      Object {
        "id": "urn:subject:36",
        "name": "Transport og logistikk Vg2",
        "nodeId": "102782",
        "text": "Transport og logistikk Vg2",
        "url": "/subjects/subject:36",
      },
    ],
  },
  Object {
    "name": "studiespesialiserende",
    "subjects": Array [
      Object {
        "id": "urn:subject:42",
        "name": "Biologi 1",
        "nodeId": "52234",
        "text": "Biologi 1",
        "url": "/subjects/subject:42",
      },
      Object {
        "id": "urn:subject:17",
        "lang": "en",
        "name": "Engelskspråklig litteratur og kultur",
        "nodeId": "71085",
        "text": "Engelskspråklig litteratur og kultur",
        "url": "/subjects/subject:17",
      },
      Object {
        "id": "urn:subject:27",
        "lang": "en",
        "name": "Internasjonal engelsk",
        "nodeId": "56850",
        "text": "Internasjonal engelsk",
        "url": "/subjects/subject:27",
      },
      Object {
        "id": "urn:subject:2",
        "name": "Kinesisk 1",
        "nodeId": "127756",
        "text": "Kinesisk 1",
        "url": "/subjects/subject:2",
      },
      Object {
        "id": "urn:subject:2",
        "name": "Kinesisk 2",
        "nodeId": "138654",
        "text": "Kinesisk 2",
        "url": "/subjects/subject:2",
      },
      Object {
        "id": "urn:subject:18",
        "name": "Kommunikasjon og kultur 1",
        "nodeId": "6118",
        "shortname": "KK1",
        "text": "Kommunikasjon og kultur 1",
        "url": "/subjects/subject:18",
      },
      Object {
        "id": "urn:subject:18",
        "name": "Kommunikasjon og kultur 2",
        "nodeId": "2602",
        "shortname": "KK2",
        "text": "Kommunikasjon og kultur 2",
        "url": "/subjects/subject:18",
      },
      Object {
        "id": "urn:subject:18",
        "name": "Kommunikasjon og kultur 3",
        "nodeId": "2603",
        "shortname": "KK3",
        "text": "Kommunikasjon og kultur 3",
        "url": "/subjects/subject:18",
      },
      Object {
        "id": "urn:subject:7",
        "name": "Markedsføring og ledelse 1",
        "nodeId": "52293",
        "text": "Markedsføring og ledelse 1",
        "url": "/subjects/subject:7",
      },
      Object {
        "id": "urn:subject:32",
        "name": "Matematikk R1",
        "nodeId": "57933",
        "text": "Matematikk R1",
        "url": "/subjects/subject:32",
      },
      Object {
        "id": "urn:subject:32",
        "name": "Matematikk R2",
        "nodeId": "98361",
        "text": "Matematikk R2",
        "url": "/subjects/subject:32",
      },
      Object {
        "id": "urn:subject:31",
        "name": "Matematikk S1",
        "nodeId": "57934",
        "text": "Matematikk S1",
        "url": "/subjects/subject:31",
      },
      Object {
        "id": "urn:subject:31",
        "name": "Matematikk S2",
        "nodeId": "98366",
        "text": "Matematikk S2",
        "url": "/subjects/subject:31",
      },
      Object {
        "id": "urn:subject:14",
        "name": "Medie- og informasjonskunnskap",
        "text": "Medie- og informasjonskunnskap",
        "url": "/subjects/subject:14",
      },
      Object {
        "beta": true,
        "id": "urn:subject:1",
        "name": "Medieuttrykk og mediesamfunnet",
        "nodeId": "156500",
        "text": "Medieuttrykk og mediesamfunnet",
        "url": "/subjects/subject:1",
      },
      Object {
        "id": "urn:subject:23",
        "lang": "en",
        "name": "Samfunnsfaglig engelsk",
        "nodeId": "71082",
        "text": "Samfunnsfaglig engelsk",
        "url": "/subjects/subject:23",
      },
      Object {
        "beta": true,
        "id": "urn:subject:43",
        "name": "Sosiologi og sosialantropologi",
        "nodeId": "185103",
        "text": "Sosiologi og sosialantropologi",
        "url": "/subjects/subject:43",
      },
      Object {
        "id": "urn:subject:8",
        "name": "Tysk 1",
        "nodeId": "137416",
        "text": "Tysk 1",
        "url": "/subjects/subject:8",
      },
      Object {
        "id": "urn:subject:8",
        "name": "Tysk 2",
        "nodeId": "138655",
        "text": "Tysk 2",
        "url": "/subjects/subject:8",
      },
    ],
  },
]
`;
